plutus-core/plutus-core/src/PlutusCore/Constant/Typed.hs

```diff
@@ -521,6 +526,46 @@ instance (HasConstantIn uni term, GShow uni, GEq uni, uni `Contains` a) =>
 -- | A constraint for \"@a@ is a 'KnownType' by means of being included in @UniOf term@\".
 type KnownBuiltinType term a = KnownBuiltinTypeIn (UniOf term) term a
 
+{- Note [Performance of KnownTypeIn instances]
+Even though we don't use 'makeKnown' and 'readKnown' directly over concrete types, it's still
+beneficial to inline them, because otherwise GHC compiles each of them to two definitions
+(one calling the other) for some reason.
+
+Similarly, we inline implementations of 'knownTypeAst' just to get tidier Core.
+
+Some 'readKnown' implementations require inserting a call to 'oneShot'. E.g. if 'oneShot' is not
+used in 'readKnownConstant' then 'GHC pulls @gshow uniExp@ out of the 'Nothing' branch, thus
+allocating a thunk of type 'String' that is completely redundant whenever there's no error,
+which is the majority of cases. And putting 'oneShot' as the outermost call results in
+worse Core.
+
```

Comment:
Is there any concrete way that we can say what "worse Core" means?  If we have to look at the generated Core, how do we know if it's worse or better?  Would it generally be the case that smaller=better, for instance?  I'm wondering if there's any possibility of some automated analysis.

Reply:
> Is there any concrete way that we can say what "worse Core" means?

I've added a comment about that, but the answer is basically "no". A tool can't figure out if sharing is actually needed or not and lots of optimizations amount to figuring out if need to drop a `let` or add one.